### ISSUE TITLE
Latent-translator-based distillation across different latent spaces

### ISSUE THREAD

**Issue by wendlerc** [Wayfarer-Labs/owl-wms#17]
In order to distill general world knowledge from open-source text2video models like Wan (teacher) to OWL-WMs (student), we need a way to distill between different latent spaces. Established distillation approaches are not designed for cross-latent space distillation and either break or incur significant costs due to the need to convert latents between latent spaces.

# TODOs

**Train latent translator**
- [x] Create latent translator dataset by sampling videos and encoding them with the two encoders.
- [x] Write latent translator trainer & train.

**Distillation**
- [ ] Incorporate the translator into, e.g., the [CausVid](https://causvid.github.io) pipeline.

# Context

## Vanilla distribution matching diffusion (recap)

The goal of original distribution matching diffusion paper is to distill a base diffusion model $\epsilon_{base}(z_t, t)$ (teacher) into a 1-step generator $G_{\theta}(z_T)$ (student), with $z_T \sim N(0, I)$, and just to be clear in the original setup $Z^{(1)} = Z^{(2)} = Z$, i.e., a both teacher and student use a shared latent space. Also the student is initialized from the teacher's weights by setting.

Looking at Fig. 2 of the [paper](https://arxiv.org/abs/2311.18828) we see the entire setup summarized:

![Image](https://github.com/user-attachments/assets/a02dc6a2-ff51-4287-a4a7-cf51b58777b0)

It looks a bit cluttered, but essentially consists of three models that are used to define two gradients to update the student. The first one (right side of the figure) is an approximation of the gradient of the **distribution matching loss**. 

![](https://quicklatex.com/cache3/07/ql_36626c3cd4c6b52f88d2c6fbf0472107_l3.png)

The second one is obtained by simple supervised learning on a dataset of teacher (input, output)-pairs and a **regression loss** (left side of the figure).

### Distribution matching gradient approximation

For this part it is useful to recall the connection between diffusion models and score functions. [Kingma et al.](https://arxiv.org/abs/2107.00630) show in appendix L that given infinite data minimizing the diffusion loss results in $\nabla_{z_t} \log p(z_t) = s_{base}(z_t, t) = \frac{\epsilon_{base}(z_t, t)}{\sigma_t}$, thus, for finite data we have $\nabla_{z_t} \log q(z_t) \approx s_{base}(z_t, t) = \frac{\epsilon_{base}(z_t, t)}{\sigma_t}$. In essence distribution matching gradient approximation works by replacing the intractable score functions of the real distributions with soft ones parametrized by the base model.

To see this let's expand the relevant part of the backward KL divergence from above $\log\left(\frac{p_{fake}(z)}{p_{real}(z)}\right) = \log (p_{fake}(z)) - \log (p_{real}(z)).$

Taking the gradient yields $(\nabla_{z} \log (p_{fake}(z)) - \nabla_{z} \log (p_{real}(z))) \frac{\partial G}{\partial \theta}$ (via chain rule since $z = G(z_T)$). 

Now, $\nabla_{z} \log (p_{real}(z))$ is approximated by $s_{real}(z_t, t)$, a score parametrized by a frozen version of the teacher model, and $\nabla_{z} \log (p_{fake}(z))$ by $s_{fake}(z_t, t)$ another copy of the teacher model that is updated throughout the distillation to track $G$'s (the student's) distribution. Also the expectation is taken over sampled student outputs that are noised by sampling random timesteps and using the teacher's forward process.

![Image](https://github.com/user-attachments/assets/a404afa0-3f86-46d5-abd6-7707cb34d8f1).

## Cross-latent space distribution matching diffusion

Now that we have laid out the key components of DMD it is straightforward to see what needs to be adapted for the case in which teacher and student don't share the same latent space. 

Let's call latents in the latent space of the teacher model $z^{(1)} \in Z^{(1)}$ and latents in the latent space of the student model $z^{(2)} \in Z^{(2)}$. Each latent space comes with a encoder, decoder pair $E^{(i)}, D^{(i)}$ with $E^{(i)}(x) = z^{(i)}$ and $D^{(i)}(z^{(i)}) = x$ for $i = 1, 2$ that encode/decode data points $x \in X$, e.g., videos/images/etc.. Further, let's call our latent translator model $T^{(2,1)}: Z^{(2)} \to Z^{(1)}: z^{(2)} \mapsto z^{(1)}$.

### Initialization 

Because the student and the teacher are now in separate latent spaces and probably also have separate architectures in our case (?) the student can no longer be directly initialized via the teacher's weights. **There are many ways to do this, curious to hear your inputs**. 

### Adaptation of the generator

For the sake of computing both the DMD gradient approximation and the supervised loss the $Z^{(2)}$-generator $z^{(2)} = G(z^{(2)}_T)$ can be turned into a $Z^{(1)}$-generator by composing it with the translator $z^{(1)} = T^{(1,2)}(G(z^{(2)}_T))$. Doing so breaks none of the math and skips the costly decoder-encoder-based translation step $z^{(1)} = E^{(1)}(D^{(2)}(z^{(2)}))$.

**Notes:** This formulation is quite nice and flexible. It allows us to choose whether we want to update the translator during the distillation and also it would allow to switch between our cheap and the costly decoder-encoder-based translator. 

### Generality of the recipe

I checked and the same arguments should hold for [CausVid](https://causvid.github.io).

# Related issues

https://github.com/Wayfarer-Labs/owl-wms/issues/13 provides an alternative solution to this problem that does not require translators and instead leverages reinforcement learning.

**Comment by wendlerc** [comment]
## Update
- I created pipelines for creating latent pair datasets: https://github.com/wendlerc/latent2latent/blob/main/generate_paired_dataset_async.py
- I wrote a simple training script: https://github.com/wendlerc/latent2latent/blob/main/train.py

## Thoughts
Today we had some signs of life of the translators actually doing something nontrivial (see attached). However, the results seem still far from usable.

![Image](https://github.com/user-attachments/assets/dcd7a3ac-bf2d-48db-b48c-b0df8a8da432)

### Connection between translator training and one-step generator training

The translator's goal is similar to a one-step generator's goal in that it also wants to create samples that look realistic. Thus, maybe the training can be sped up / improved by performing a DMD-type update.

$s_{real}$... should be maybe just WAN diffusion backbone (from now on WAN).
$s_{fake}$... should be WAN tracking the translators outputs.

Voila (?). Are these DMD gradients richer than just minimizing MSE: $\frac{1}{n} \|\text{translator}(z) - \text{encoder}(\text{decoder}(z)\|^2$?

**Comment by wendlerc** [comment]
Scaling from 2.2 M trainable params to 7.6 M and bit larger dataset and more steps:

![Image](https://github.com/user-attachments/assets/86929662-ec4c-49e6-a899-88dbfe18dbfd)

Loss curves: 

<img width="2150" height="1184" alt="Image" src="https://github.com/user-attachments/assets/2e0fbc1a-ee05-4063-8073-59ad9252f638" />

Final loss of the new model: 0.442
